src/expr/src/table_function/generate_series.rs

```diff
@@ -0,0 +1,217 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use itertools::multizip;
+use risingwave_common::array::{
+    Array, ArrayBuilder, ArrayImpl, ArrayRef, DataChunk, I32Array, IntervalArray,
+    NaiveDateTimeArray,
+};
+use risingwave_common::ensure;
+use risingwave_common::types::{CheckedAdd, Scalar, ScalarRef};
+use risingwave_common::util::chunk_coalesce::DEFAULT_CHUNK_BUFFER_SIZE;
+
+use super::*;
+use crate::ExprError;
+
+#[derive(Debug)]
+pub struct GenerateSeries<T: Array, S: Array> {
+    start: BoxedExpression,
+    stop: BoxedExpression,
+    step: BoxedExpression,
+    _phantom: std::marker::PhantomData<(T, S)>,
+}
+
+impl<T: Array, S: Array> GenerateSeries<T, S>
+where
+    T::OwnedItem: for<'a> PartialOrd<T::RefItem<'a>>,
+    T::OwnedItem: for<'a> CheckedAdd<S::RefItem<'a>, Output = T::OwnedItem>,
+{
+    fn new(start: BoxedExpression, stop: BoxedExpression, step: BoxedExpression) -> Self {
+        Self {
+            start,
+            stop,
+            step,
+            _phantom: Default::default(),
+        }
+    }
+
+    fn eval_row(
```

Comment:
I originally proposed 
```rust
pub trait TableFunction: {
    fn return_type(&self) -> DataType;

    fn eval_row(&self, input: &Row) -> Result<ArrayRef>;
}
```
But @TennyZhuang suggets accepting a Chunk and outputs Vec of Array. I'm not sure which is better 🤔

Reply:
slightly +1 for your proposal

Reply:
But `eval_row` is also needed, we can add the vectorized version later.

Reply:
For this PR, I have no problem.

Reply:
Created another issue for vectorized version. I will change the current version to row based 
 https://github.com/singularity-data/risingwave/issues/3892

Reply:
> It seems that both `generate_series` and `unnest` follows a similar pattern: accept DataChunk then do `eval_row` row-by-row. Could this be extracted into the TableFunction trait?

Actually normal exprs also follow this pattern? (I just expanded the macro by hand 😇) The eval is row-by-row, but the input arg eval is vectorized. Also `eval` doesn't call `eval_row`, but uses macro, since the type of the evaluation logic cannot be captured by trait.

https://github.com/singularity-data/risingwave/blob/7a86c4d3ad508c00d8a18ce58fcf7252a172d90b/src/expr/src/expr/template.rs#L49-L79

So I think it is fine to just keep current status in this PR?

Reply:
LGTM